### PR TITLE
Auto Save All Buffers After A Delay

### PR DESCRIPTION
**Pull request by miggs597**
The feature I would like to introduce with this pull request is the ability to auto save one's work after idling. I've found it to be a great editing experience, especially while working with a stack that offers hot module reloading.

My implementation works similarly to `idle_timer`, that is once my `save_delay_timer` has reached zero an `EditorEvent::SaveDelayTimer` is returned causing `EditorView::handle_event` to handle that timeout and save the buffers.

Any feedback is greatly appreciated.

### CODE REVIEW

**Comment by woojiq** [comment]
Sorry, I don't understand the use case if helix already has `auto-save`:
> Enable automatic saving on the focus moving away from Helix.

Can you explain more about  hot reloading usage? Do you want to automatically save every second or what?

**Comment by kirawi** [comment]
Resolves https://github.com/helix-editor/helix/issues/3451

**Comment by miggs597** [comment]
> Sorry, I don't understand the use case if helix already has `auto-save`:
> 
> > Enable automatic saving on the focus moving away from Helix.
> 
> Can you explain more about hot reloading usage? Do you want to automatically save every second or what?

Yes, helix does already have auto-save. However if enabled it's only triggered by changing focus away from helix. This change would allow for buffers to get saved after a certain amount of time has elapsed without any new input. 

Before I was using vscode as my text editor and got used to it there. https://code.visualstudio.com/docs/editor/codebasics#_save-auto-save

**Comment by miggs597** [comment]
> This is definitely useful but it shouldn't be implemented with the idle timeout we will remove that soonish (once all items in the tracking issue #9629 are done)

Can you point me in the direction of some code samples that use the event system. I can work on updating my pr to use it instead of the current ad-hoc method.

**Comment by kirawi** [comment]
https://github.com/helix-editor/helix/tree/master/helix-term/src/handlers contain them

**Comment by miggs597** [comment]
Moving to a draft for now. I've started working on porting the idea over to the event system.

**Comment by miggs597** [comment]
> This is definitely useful but it shouldn't be implemented with the idle timeout we will remove that soonish (once all items in the tracking issue #9629 are done)

@pascalkuthe I realized that my initial pr description wasn't clear. I am not piggybacking off of idle timeout, however I do depend on the `event_loop_until_idle`. That's where it is being checked if the save timeout has been reached. 

Is `Application::event_loop_until_idle` also being ported over to the event system, or is that code going to stay as is?
I'm trying to figure out from where it would be most appropriate to dispatch an auto save event after a duration has expired.

**Comment by pascalkuthe** [comment]
All of that is going away. Eberything is getting its own async handler with its own timout.

Moving around should not prevent autosave. You should be registering documentdidchange handler and start a (fairly long like 5s) timeout there that will trigger autosave when finished. Like kirawi said look at how the signature help handler is implemented (just that you don't care about mode and selection changes)

**Comment by miggs597** [comment]
Initial port to the event system, implementation is currently half-baked. I need to iron out the user experience, and to get it working consistently.

File: helix-term/src/handlers/auto_save.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+#[derive(Debug)]
+enum State {
+    Closed,
+}
+
+#[derive(Debug)]
+pub(super) struct AutoSaveHandler {
+    trigger: Option<AutoSaveInvoked>,
```
Comment (pascalkuthe) [inline helix-term/src/handlers/auto_save.rs:30]: There is no need to keep the trigger around , you don't actually use this

File: helix-term/src/handlers/auto_save.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        editor,
+        scroll: Some(0),
+        jobs: &mut Jobs::new(),
+    };
+
+    if let Err(e) = commands::typed::write_all_impl(context, false, false) {
+        context.editor.set_error(format!("{}", e));
+    }
+
+    if let Err(e) = context.block_try_flush_writes() {
```
Comment (pascalkuthe) [inline helix-term/src/handlers/auto_save.rs:87]: I don't think thsi should be here, this is only intended for shutdown. We don't want to freeze the editor while writing to disk

**Review by pascalkuthe — COMMENTED** [review]
This is actually already a really good implementation. I have some minor comments but apart from that this seems reads to merge actually

File: helix-view/src/editor.rs [hunk 3]
```diff
@@ -840,7 +856,9 @@ impl Default for Config {
             auto_completion: true,
             auto_format: true,
             auto_save: false,
+            save_style: SaveStyle::FocusLost,
             idle_timeout: Duration::from_millis(250),
+            save_delay_timeout: Duration::from_millis(1000),
```
Comment (pascalkuthe) [inline helix-view/src/editor.rs:861]: Maybe the default should be longer like 5s?

Comment (Banzobotic) [reply helix-view/src/editor.rs:861]: For me the primary benefit of auto saving after a delay is updating lints from LSPs such as rust-analyzer. IMO one second is the best default for this use case. This also seems to be main use case mentioned on #3451. 1s is also the default in vscode and it works well there.

**Comment by pascalkuthe** [comment]
closing in favor of #10899

**Pull request closed without merging**